### PR TITLE
test (azure-iot-device): fix provisioning pipeline tests to only test pipeline objects  and not…

### PR DESCRIPTION
**Pull request by BertKleewein**
… hang at end

### CODE REVIEW

File: azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py [hunk 1]
```diff
@@ -327,226 +312,129 @@ def test_op_fail(self, mocker, pipeline, arbitrary_exception):
...
 @pytest.mark.describe("ProvisioningPipeline - .register()")
 class TestSendRegister(object):
-    @pytest.mark.it("Request calls publish on provider")
-    def test_send_register_request_calls_publish_on_provider(
-        self, mocker, mock_provisioning_pipeline, input_security_client, mock_mqtt_transport
-    ):
-        mock_init_uuid = mocker.patch(
-            "azure.iot.device.common.pipeline.pipeline_stages_base.uuid.uuid4"
+    @pytest.mark.it("Runs a RegisterOperation on the pipeline")
+    def test_runs_op(self, pipeline, mocker):
```
Comment (cartertinney) [inline azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py]: Validate the request payload and the registration id on this operation

File: azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py [hunk 2]
```diff
@@ -327,226 +312,129 @@ def test_op_fail(self, mocker, pipeline, arbitrary_exception):
...
+        cb = mocker.MagicMock()
+        pipeline.register(callback=cb)
+        assert pipeline._pipeline.run_op.call_count == 1
+        assert isinstance(
+            pipeline._pipeline.run_op.call_args[0][0], pipeline_ops_provisioning.RegisterOperation
```
Comment (cartertinney) [inline azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py]: Also validate the contents of the operation: the registration id and request payload

File: azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py [hunk 3]
```diff
@@ -327,226 +312,129 @@ def test_op_fail(self, mocker, pipeline, arbitrary_exception):
...
-        if isinstance(input_security_client, X509SecurityClient):
-            assert mock_mqtt_transport.connect.call_args[1]["password"] is None
-        else:
-            assert mock_mqtt_transport.connect.call_args[1]["password"] is not None
-            assert_for_symmetric_key(mock_mqtt_transport.connect.call_args[1]["password"])
 
-        fake_publish_topic = "$dps/registrations/PUT/iotdps-register/?$rid={}".format(
-            fake_request_id
-        )
+    @pytest.mark.it("Triggers the callback upon successful completion of the RegisterOperation")
```
Comment (cartertinney) [inline azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py]: Probably could do to mention that the registration result is returned

File: azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py [hunk 4]
```diff
@@ -327,226 +312,129 @@ def test_op_fail(self, mocker, pipeline, arbitrary_exception):
...
+        assert cb.call_count == 1
+        assert cb.call_args == mocker.call(error=arbitrary_exception, result=None)
 
-        # verify that our connected callback was called and verify that we published the event
-        mock_provisioning_pipeline.on_connected.assert_called_once_with("connected")
 
-        fake_publish_topic = "$dps/registrations/PUT/iotdps-register/?$rid={}".format(
-            fake_request_id
+@pytest.mark.describe("ProvisioningPipeline - .disconnect()")
+class TestDisconnect(object):
```
Comment (cartertinney) [inline azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py]: >TestDisconnect [](start = 6, length = 14)

There was already a test for this method above

File: azure-iot-device/azure/iot/device/provisioning/pipeline/provisioning_pipeline.py [hunk 5]
```diff
@@ -207,7 +207,7 @@ def on_complete(op, error):
             if error:
                 callback(error=error, result=None)
             else:
-                callback(result=op.registration_result)
+                callback(error=None, result=op.registration_result)
```
Comment (cartertinney) [inline azure-iot-device/azure/iot/device/provisioning/pipeline/provisioning_pipeline.py]: I don't really understand the need for this change

**Comment by cartertinney** [comment]
>             pipeline_ops_base.EnableFeatureOperation(feature_name=None, callback=pipeline_callback)

This is not part of this PR, but this is.... weird.

---
Refers to: azure-iot-device/azure/iot/device/provisioning/pipeline/provisioning_pipeline.py:187 in 89b6e60. [](commit_id = 89b6e60c632ef5644636aca6770e3bcd72f175a2, deletion_comment = False)

File: azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py [hunk 6]
```diff
@@ -327,226 +312,129 @@ def test_op_fail(self, mocker, pipeline, arbitrary_exception):
...
+@pytest.mark.describe("ProvisioningPipeline - .enable_responses()")
+class TestEnable(object):
+    @pytest.mark.it("Marks the feature as enabled")
+    def test_mark_feature_enabled(self, pipeline, mocker):
+        assert not pipeline.responses_enabled[feature]
+        pipeline.enable_responses(callback=mocker.MagicMock())
+        assert pipeline.responses_enabled[feature]
 
-        mock_mqtt_transport.disconnect.assert_called_once_with()
+    @pytest.mark.it("Runs a EnableFeatureOperation on the pipeline")
```
Comment (cartertinney) [inline azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py]: >EnableFeatureOperation [](start = 28, length = 22)

We should also validate the feature name on the operation, as we do in the hub's mqtt pipeline.

Except, as pointed out in another comment, for some reason it doesn't have one here? We probably shouldn't be enabling a 'None' feature, and just always taking that to be the single item we can enable.

File: azure-iot-device/tests/provisioning/pipeline/test_pipeline_stages_provisioning.py [hunk 7]
```diff
@@ -50,6 +50,14 @@ def apply_fake_pipeline_thread(fake_pipeline_thread):
     pass
 
 
+# automatically mock the transport for all tests in this file.
+@pytest.fixture(autouse=True)
```
Comment (olivakar) [inline azure-iot-device/tests/provisioning/pipeline/test_pipeline_stages_provisioning.py]: >pytest [](start = 1, length = 6)

was something failing in horton without this ?

**Comment by olivakar** [comment]
> class TestProvisioningPipelineDisconnect(object):

just rename to TestDisconnect and delete the duplicate ones later.

---
Refers to: azure-iot-device/tests/provisioning/pipeline/test_provisioning_pipeline.py:275 in 89b6e60. [](commit_id = 89b6e60c632ef5644636aca6770e3bcd72f175a2, deletion_comment = False)

File: azure-iot-device/tests/provisioning/pipeline/test_pipeline_stages_provisioning.py (hunk 7 above)
Comment (BertKleewein) [reply azure-iot-device/tests/provisioning/pipeline/test_pipeline_stages_provisioning.py]: It turns out this was unnecessary.  I did this as an experiment and never pulled it out.

---
In reply to: [394571019](https://github.com/Azure/azure-iot-sdk-python/pull/499#discussion_r394571019) [](ancestors = 394571019)

File: azure-iot-device/azure/iot/device/provisioning/pipeline/provisioning_pipeline.py (hunk 5 above)
Comment (BertKleewein) [reply azure-iot-device/azure/iot/device/provisioning/pipeline/provisioning_pipeline.py]: I did it for test consistency, but I agree that it's not necessary

---
In reply to: [394562682](https://github.com/Azure/azure-iot-sdk-python/pull/499#discussion_r394562682) [](ancestors = 394562682)

**Pull request merged**